### ISSUE TITLE
Unified Alerting: nil dereference in AlertManager when creating silence as anonymous

### ISSUE THREAD

**Issue by Sameth** [grafana/grafana#47155]
**What happened**:

Attempted to create silence with Unified Alerting. Received an internal server error. Stacktrace from the logs towards the end of the issue.

Background info: The grafana instance was running with `[auth.anonymous]` enabled, I was using the anonymous "user".

**What you expected to happen**:

Silence is created successfully.

**How to reproduce it (as minimally and precisely as possible)**:

- Run Grafana with `[auth.anonymous] enabled=true` in `grafana.ini`
  - Unified alerting should be enabled (should be the default now)
- Create silence in the Alerting tab

**Anything else we need to know?**:

- Failing line: https://github.com/prometheus/alertmanager/blob/v0.23.0/api/v2/compat.go#L87
- CreatedBy field seems to be coming from here (not 100% sure, feel free to verify): https://github.com/grafana/grafana/blob/release-8.4.4/public/app/features/alerting/unified/components/silences/SilencesEditor.tsx#L84
  - grafanaBootData.user.name was verified to be empty in the Developer Tools console in Chromium
- Stacktrace:

```
logger=context t=2022-03-24T15:10:20.3+0000 lvl=eror msg="Request error" error="runtime error: invalid memory address or nil pointer dereference" stack="/usr/local/go/src/runtime/panic.go:221 (0x44df86)
/usr/local/go/src/runtime/signal_unix.go:735 (0x44df56)
/opt/drone/gomodcache/[github.com/prometheus/alertmanager@v0.23.1-0.20211116083607-e2a10119aaf7/api/v2/compat.go:87](http://github.com/prometheus/alertmanager@v0.23.1-0.20211116083607-e2a10119aaf7/api/v2/compat.go:87) (0x173b8a9)
/drone/src/pkg/services/ngalert/notifier/silences.go:78 (0x174c267)
/drone/src/pkg/services/ngalert/api/api_alertmanager.go:137 (0x1c57a03)
/drone/src/pkg/services/ngalert/api/forked_am.go:57 (0x1c62942)
/drone/src/pkg/services/ngalert/api/generated_base_api_alertmanager.go:57 (0x1c6353a)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/services/ngalert/metrics/ngalert.go:286 (0x172b0f6)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0xe2d39a)
/drone/src/pkg/web/macaron.go:155 (0xe2d38c)
/usr/local/go/src/net/http/server.go:2047 (0x6d2d4e)
/drone/src/pkg/middleware/csp.go:22 (0x1720832)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0xe2d39a)
/drone/src/pkg/web/macaron.go:155 (0xe2d38c)
/usr/local/go/src/net/http/server.go:2047 (0x6d2d4e)
/drone/src/pkg/web/render.go:44 (0xe28146)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0xe2d39a)
/drone/src/pkg/web/macaron.go:155 (0xe2d38c)
/usr/local/go/src/net/http/server.go:2047 (0x6d2d4e)
/drone/src/pkg/middleware/csrf.go:21 (0x1720c4f)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0x1723b4a)
/drone/src/pkg/middleware/recovery.go:168 (0x1723b41)
/drone/src/pkg/web/context.go:32 (0xe252fb)
/drone/src/pkg/web/inject.go:139 (0xe261f4)
/drone/src/pkg/web/inject.go:117 (0xe2611c)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0x1721904)
/drone/src/pkg/middleware/logger.go:33 (0x17218ed)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0x172482f)
/drone/src/pkg/middleware/request_metrics.go:60 (0x172481e)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/context.go:62 (0x1725fb9)
/drone/src/pkg/middleware/request_tracing.go:60 (0x1725fa8)
/usr/local/go/src/reflect/value.go:556 (0x4df704)
/usr/local/go/src/reflect/value.go:339 (0x4dec84)
/drone/src/pkg/web/inject.go:159 (0xe26486)
/drone/src/pkg/web/inject.go:119 (0xe260d4)
/drone/src/pkg/web/context.go:67 (0xe2540b)
/drone/src/pkg/web/router.go:156 (0xe299bd)
/drone/src/pkg/web/router.go:223 (0xe2a651)
/drone/src/pkg/web/macaron.go:189 (0xe27a90)
/usr/local/go/src/net/http/server.go:2879 (0x6d62ba)
/usr/local/go/src/net/http/server.go:1930 (0x6d1e27)
/usr/local/go/src/runtime/asm_amd64.s:1581 (0x46bf00)
```

**Environment**:
- Grafana version: grafana/grafana-oss:8.4.4 docker image
- Data source type & version: N/A (Alerting doesn't have DataSources)
- OS Grafana is installed on: Ubuntu 20.04
- User OS & Browser: Ubuntu 20.04, Chromium (~up-to-date)
- Grafana plugins: N/A

**Comment by armandgrillet** [comment]
I believe https://github.com/grafana/grafana/pull/46997 fixes the issue (this has been backported to 8.4 and 8.5), can you confirm?

**Comment by aw-fbox** [comment]
Having this same issue, confirmed still exists in "latest" and 8.4.5.

**Comment by Sameth** [comment]
TBH, I don't think anything short of fix in Grafana code will really resolve this. The `user` field is empty when you use Grafana in anonymous mode, and translates to `nil` before being passed to the AlertManager code (when you try to create the silence).

The fix will need to somehow address this. E.g. by adding a field for CreatedBy when you run Grafana in anonymous mode (better) or populating CreatedBy with e.g. `anonymous` (worse).

**Comment by alexweav** [comment]
Can we validate that the fix above covers this?

**Comment by alexweav** [comment]
I followed the reproduction steps in the current `main` and it appears to be fixed, after the PR linked above. Closing. Feel free to re-open if there is a different path to reproducing this.

**Comment by OscarStockill** [comment]
I am having a very similar issue with 9.1.5.
Unified alerting
[auth.anonymous] = true

Attempting to save while not logged in gives this error:
<img width="608" alt="silence_save_failed" src="https://user-images.githubusercontent.com/46621436/192416448-51800732-45fe-4ced-b937-54d183af144d.png">

Currently only logged in users (Admin and Editor roles) can create them successfully.

**Comment by armandgrillet** [comment]
Should be fixed in 9.2 due to https://github.com/grafana/grafana/pull/55952
